Add Relation DETR

xiuqhou opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

This PR adds Relation-DETR as introduced in [Relation DETR: Exploring Explicit Position Relation Prior for Object Detection](https://arxiv.org/abs/2407.11699). Checkpoint for Relation-DETR (ResNet50) converted from original repo https://github.com/xiuqhou/Relation-DETR has been uploaded to https://huggingface.co/xiuqhou/relation-detr-resnet50

Related issues in original repo: 
https://github.com/xiuqhou/Relation-DETR/issues/25
https://github.com/xiuqhou/Relation-DETR/issues/21

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


TODO:
- [x] Make more checkpoints with Swin-L and Focal-L backbones available on HF.
- [x] Update the document about Relation-DETR.

## Who can review?

@amyeroberts @qubvel


<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

xiuqhou committed:
Add RelationDETR

xiuqhou committed:
Align Relation-DETR with official implementation

xiuqhou committed:
Add ImageProcessor for Relation-DETR

xiuqhou committed:
Fix forward errors in Relation-DETR

xiuqhou committed:
Align RelationDetrLoss with official implementation

xiuqhou committed:
Fix tuple sequence in Relation-DETR

xiuqhou committed:
Fix data types in Relation-DETR

xiuqhou committed:
Rename def_relation_detr to relation_detr

xiuqhou committed:
Add image_processing_relation_detr tester

xiuqhou committed:
Make sure return equivalence of RelationDETR

xiuqhou committed:
Fix types of auxiliary_outputs or relation_detr

xiuqhou committed:
Fix logits and pred_boxes for dn_out

xiuqhou committed:
Fix out of range of selection in relation-detr

xiuqhou committed:
Fix initialization in RelationDETR

xiuqhou committed:
Integrate RelationDETR with transformers

xiuqhou committed:
Refactor image processor of relation-detr

xiuqhou committed:
Remove _tied_weights_keys in RelationDETR

xiuqhou committed:
Add test_modeling relation_detr

xiuqhou committed:
Correct code style

xiuqhou committed:
Style of code to pass fix-copies

xiuqhou committed:
Doc update for RelationDetrConfig

xiuqhou committed:
Code quality

xiuqhou committed:
Remove trailing whitespace

xiuqhou committed:
Remove unused `max_position_embeddings`

xiuqhou committed:
Refactor config attribute

xiuqhou committed:
Add model_doc for relation_detr

xiuqhou committed:
Update integration tests for relation-detr

xiuqhou committed:
Fix level_embed initialization of relation-detr

xiuqhou committed:
Code style  of relation-detr

qubvel commented:
Hi @xiuqhou! Congratulations on the paper, awesome work! And thanks for working on transformers implementation! Feel free to ping me when it's ready for review or if you have any questions!

xiuqhou committed:
Add RelationDetrImageProcessorFast

xiuqhou committed:
Update import structure for relation-detr

xiuqhou committed:
Fix __init__ for image_processor_fast

xiuqhou committed:
Add format postprocess of relation-detr for backbone

xiuqhou committed:
Add post-process for RelationDetr backbone

qubvel commented on docs/source/en/model_doc/relation_detr.md:
```diff
@@ -0,0 +1,102 @@
+
+<small> Performance comparison between Relation-DETR and other DETR methods. Taken from the <a href="https://arxiv.org/abs/2407.11699">original paper.</a> </small>
+
+The model version was contributed by [rafaelpadilla](https://huggingface.co/rafaelpadilla) and [xiuqhou](https://github.com/xiuqhou). The original code can be found [here](https://github.com/xiuqhou/Relation-DETR/).
```
Not sure if `rafaelpadilla` is relevant here

qubvel commented on src/transformers/loss/loss_relation_detr.py:
```diff
@@ -0,0 +1,429 @@
+import copy
```
Please Copyright add header

qubvel commented on src/transformers/loss/loss_relation_detr.py:
```diff
@@ -0,0 +1,429 @@
+        return giou_cost
+
+    @torch.no_grad()
+    def calculate_cost(self, **input_dict):
```
Do we need this to be unpacked with `**` in signature?

qubvel commented on src/transformers/loss/loss_relation_detr.py:
```diff
@@ -0,0 +1,429 @@
+        pred_masks: Optional[Tensor] = None,
+        gt_masks: Optional[Tensor] = None,
+    ):
+        c = self.calculate_cost(
```
Please avoid one-letter variables such as `c`

qubvel commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -550,6 +551,7 @@
         ("recurrent_gemma", "RecurrentGemma"),
         ("reformer", "Reformer"),
         ("regnet", "RegNet"),
+        ("relation_detr", "Relation DETR"),
```
```suggestion
        ("relation_detr", "RelationDETR"),
```

xiuqhou committed:
Fix contributor in model_doc of RelationDETR

qubvel commented on src/transformers/models/relation_detr/__init__.py:
Thanks for following the new init format!

xiuqhou committed:
Add Copyright header in loss_relation_detr

qubvel commented on src/transformers/models/relation_detr/configuration_relation_detr.py:
```diff
@@ -0,0 +1,310 @@
+                backbone_kwargs["output_stride"] = 16
+            backbone_kwargs["out_indices"] = [2, 3, 4] if num_feature_levels > 1 else [4]
+            backbone_kwargs["in_chans"] = num_channels
+        # Backwards compatibility
```
We don't need backward compatibility for new added models

qubvel commented on src/transformers/models/relation_detr/configuration_relation_detr.py:
```diff
@@ -0,0 +1,310 @@
+        return self.d_model
+
+
+__all__ = ["RelationDetrConfig"]
```
Thanks for adding `__all__`!

qubvel commented on src/transformers/models/relation_detr/configuration_relation_detr.py:
```diff
@@ -0,0 +1,310 @@
+# coding=utf-8
+# Copyright 2022 SenseTime and The HuggingFace Inc. team. All rights reserved.
```
2024? is SenseTime relevant?

qubvel commented on src/transformers/models/relation_detr/image_processing_relation_detr.py:
```diff
@@ -0,0 +1,1654 @@
+        ]
+
+    @classmethod
+    # Copied from transformers.models.detr.image_processing_detr.DetrImageProcessor.from_dict with Detr->RelationDetr
+    def from_dict(cls, image_processor_dict: Dict[str, Any], **kwargs):
+        """
+        Overrides the `from_dict` method from the base class to make sure parameters are updated if image processor is
+        created using from_dict and kwargs e.g. `RelationDetrImageProcessor.from_pretrained(checkpoint, size=600,
+        max_size=800)`
+        """
+        image_processor_dict = image_processor_dict.copy()
+        if "max_size" in kwargs:
+            image_processor_dict["max_size"] = kwargs.pop("max_size")
+        if "pad_and_return_pixel_mask" in kwargs:
+            image_processor_dict["pad_and_return_pixel_mask"] = kwargs.pop("pad_and_return_pixel_mask")
+        return super().from_dict(image_processor_dict, **kwargs)
```
This is for backward compat, we can remove it for new model

qubvel commented on src/transformers/models/relation_detr/image_processing_relation_detr.py:
```diff
@@ -0,0 +1,1654 @@
+        return encoded_inputs
+
+    # POSTPROCESSING METHODS - TODO: add support for other frameworks
+    def post_process(self, outputs, target_sizes):
```
`post_process` is deprecated, lets remove it for a new model

qubvel commented on src/transformers/models/relation_detr/image_processing_relation_detr.py:
```diff
@@ -0,0 +1,1654 @@
+            boxes = boxes * scale_fct[:, None, :]
+
+        results = []
+        for s, l, b in zip(scores, labels, boxes):
```
please avoid one-letter vars, despite it came from detr :)

qubvel commented on src/transformers/models/relation_detr/image_processing_relation_detr_fast.py:
```diff
@@ -0,0 +1,1075 @@
+        boxes = center_to_corners_format(out_bbox)
+        boxes = torch.gather(boxes, 1, topk_boxes.unsqueeze(-1).repeat(1, 1, 4))
+
+        # and from relative [0, 1] to absolute [0, height] coordinates
```
```suggestion
        # Convert from relative [0, 1] to absolute [0, height] coordinates
```

qubvel commented on src/transformers/loss/loss_relation_detr.py:
```diff
@@ -0,0 +1,429 @@
+
+        cost = class_cost + bbox_cost + giou_cost
+
+        if input_dict.get("pred_masks", None) is not None and input_dict.get("gt_masks", None) is not None:
+            cost_mask, cost_dice = self.calculate_mask_cost(**input_dict)
+            cost_mask = cost_mask * self.cost_mask
+            cost_dice = cost_dice * self.cost_dice
+
+            cost += cost_mask + cost_dice
+
```
Do we support masks with this model?

qubvel commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2788 @@
+    def forward(self, multi_level_feats: List[Tensor]):
+        if self.post_layer_norm:
+            # convert N C H W to N H W C
+            if all(feat.shape[1] == channel for feat, channel in zip(multi_level_feats, self.in_channels)):
```
Why do wee need `if` here?do we have both cases for the provided model?

xiuqhou replied:
This is for the FocalNet backbone. It gives outputs in NCHW format, we need to convert them to NHWC for layer norm, and them convert them back to NCHW format as output. To be clear:

```python
resnet (NCHW) -> output
swin   (NHWC) -> (NHWC to NCHW) -> output
focal  (NCHW) -> (NCHW to NHWC) -> layer_norm -> (NHWC to NCHW) -> output
```

qubvel commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2788 @@
+
+        self.config = config
+
+        # For backwards compatibility we have to use the timm library directly instead of the AutoBackbone API
+        if config.use_timm_backbone:
+            # We default to values which were previously hard-coded. This enables configurability from the config
+            # using backbone arguments, while keeping the default behavior the same.
+            requires_backends(self, ["timm"])
+            kwargs = getattr(config, "backbone_kwargs", {})
+            kwargs = {} if kwargs is None else kwargs.copy()
+            out_indices = kwargs.pop("out_indices", (2, 3, 4) if config.num_feature_levels > 1 else (4,))
+            num_channels = kwargs.pop("in_chans", config.num_channels)
+            if config.dilation:
+                kwargs["output_stride"] = kwargs.get("output_stride", 16)
+            backbone = create_model(
+                config.backbone,
+                pretrained=config.use_pretrained_backbone,
+                features_only=True,
+                out_indices=out_indices,
+                in_chans=num_channels,
+                **kwargs,
+            )
+        else:
+            backbone = load_backbone(config)
```
We don't need everything that is provided for backward compat, please cut these lines of code. It should be just `backbone = load_backbone(config)`

qubvel commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2788 @@
+    def forward(self, pixel_values: torch.Tensor):
+        # send pixel_values through the model to get list of feature maps
+        features = self.model(pixel_values) if self.config.use_timm_backbone else self.model(pixel_values).feature_maps
+        features = self.post_process(features)
```
Would be nice to have a better name, e.g. features_layer_norm (for module as well)

qubvel commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2788 @@
+        super().__init__()
+        self.in_channels = in_channels
+        self.post_layer_norm = post_layer_norm
+        if self.post_layer_norm:
```
Are there cases when `post_layer_norm` is False?

xiuqhou replied:
Yes, layer_norm is needed for focalnet backbone but not for resnet and swin backbones. 

I don't know why they have different settings, but this behavior comes from the framework detrex (https://github.com/IDEA-Research/detrex/blob/main/detrex/modeling/backbone/focalnet.py#L474-L478) and Relation-DETR simply follows it.

qubvel commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2788 @@
+    need paper, generalized to work on images.
+    """
+
+    def __init__(self, embedding_dim=64, temperature=10000, normalize=False, scale=2 * math.pi, eps=1e-6, offset=0.0):
```
Should take `config` as input param only, similar to other modules

qubvel commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2788 @@
+    return position_embedding
+
+
+def multi_scale_deformable_attention(
```
is this copied? then we should add `# Copied from`

qubvel commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2788 @@
+        return output, attention_weights
+
+
+class RelationDetrMultiheadAttention(nn.Module):
```
Is it copied?

qubvel reviewed (commented):
Hey @xiuqhou! Thanks a lot for working on the model addition! It looks very good for the first iteration, and I'm impressed with how fast you were able to make it. Great job!

I made an initial review, the main takeaways:

- There should be a conversion script from the original checkpoint to the HF-like checkpoint. Please follow the `mllama` model conversion script format and structure.
- Everywhere possible, please add `# Copied from ...` statements if a module was copied without modifications. If a module was copied and modified slightly, please add a comment `# Modified from ...`.
- Please remove all if/else statements where possible. In cases where the provided checkpoints do not offer different options (e.g., an else statement is never reachable), it should be removed.
- Avoid one-letter variables, add type hints, and follow the transformers-like docstring style.

Once again, thanks for working on this! It will be a great addition to transformers! And feel free to ask questions if anything is not clear!

xiuqhou committed:
Remove **kwargs and masks in RelationDetrHungarianMatcher

xiuqhou commented:
Thanks! I will fix them accordingly.

xiuqhou committed:
Refactor RelationDetrLoss params to RelationDetrConfig

xiuqhou committed:
Update docstring format in loss of RelationDETR

xiuqhou committed:
Update src/transformers/models/auto/configuration_auto.py

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

xiuqhou committed:
Update Relation DETR to RelationDETR in doc table

xiuqhou committed:
Update src/transformers/models/relation_detr/image_processing_relation_detr_fast.py

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

xiuqhou committed:
Remove code about backward compatability in RelationDetrConfig

xiuqhou committed:
Update comment in RelationDetrImageProcessor

xiuqhou committed:
Fix error caused by backbone_config dict

xiuqhou committed:
Fix Copyright header in RelationDETR

xiuqhou committed:
Remove backward compatability and unused params in image_processing of RelationDETR

xiuqhou committed:
Add copied from comment for RelationDETR

xiuqhou committed:
Refactor docstring and param for RelationDetr

xiuqhou committed:
Refactor RelationDetrConvEncoderPostLayerNorm

xiuqhou committed:
Add `Copied from` and `Modified from` for RelationDETR

xiuqhou committed:
Fix threshold filter in post process of RelationDetr

xiuqhou committed:
Update test case of RelationDETR

xiuqhou committed:
Use config as input param of RelationDetrSinePositionEmbedding

xiuqhou committed:
Remove position_embedding_type in RelationDetrConfig

xiuqhou committed:
Remove back compat in backbone of RelationDetr

xiuqhou committed:
Update test config of RelationDetr

xiuqhou committed:
Add more type hint and docstring for RelationDetr

xiuqhou committed:
Add convert script for resnet and swin backbone

xiuqhou committed:
Add convert script for focalnet backbone

xiuqhou committed:
Use print rather than logger for stdout

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2788 @@
+            Logits of predicted bounding boxes coordinates in the first stage.
+    """
+
+    init_reference_points: torch.FloatTensor = None
+    dec_outputs_class: torch.FloatTensor = None
+    dec_outputs_coord: torch.FloatTensor = None
```
Hi, I have a question. I noticed that other models in transformers also set returns to `None` by default, even when not optional. Should I follow the pattern and keep them as None for consistency?

qubvel replied:
Typing is not always consistent in the repo. We are trying to improve it. If possible, make it None by default only if it's required.

qubvel replied:
In case of any test or consistency check difficulties, it's okay to set it to "None" by default.

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2788 @@
+                feat = getattr(self, f"norm{idx}")(feat)
+                multi_level_feats[idx] = feat
+
+        # convert N H W C to N C H W
+        if all(feat.shape[-1] == channel for feat, channel in zip(multi_level_feats, self.in_channels)):
+            multi_level_feats = [feat.permute(0, 3, 1, 2) for feat in multi_level_feats]
```
Some timm backbones return outputs in different formats by default (e.g., NCHW for ResNet and NHWC for Swin). I didn't find a parameter to set the output format, so I added code to handle the conversion. Is there a better way to achieve this?

xiuqhou commented:
Hi @qubvel  Thanks again for your review! I’ve updated the code based on your feedback, added a new checkpoint conversion script, improved the docstring, added type hints, and included comments for `Copied from` and `Modified from`. Also, for those may not need change, I added corresponding explanations under your reviews.

qubvel commented:
Hi @xiuqhou, thanks for addressing the comments, and sorry for the delay, will review this week!

qubvel commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2871 @@
+
+        self.config = config
+        backbone = load_backbone(config)
+
+        # replace batch norm by frozen batch norm
+        with torch.no_grad():
+            replace_batch_norm(backbone)
+        self.model = backbone
```
Do we need a torch.no_grad here?

qubvel reviewed (commented):
Hi @xiuqhou, the PR looks great! I left a few more comments below, thanks a lot for interating 🤗

xiuqhou committed:
Remove no_grad context for replace_batch_norm

xiuqhou committed:
Update param comment with more descriptive names

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

xiuqhou committed:
Update param comment

xiuqhou committed:
Transpose features based on backbone_features_format

xiuqhou committed:
Update src/transformers/models/relation_detr/modeling_relation_detr.py

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

xiuqhou committed:
Update param and docstring for `backbone_features_format`

xiuqhou committed:
Remove `with_pos_embed`

xiuqhou committed:
Use backbone_features_format in RelationDetrConvEncoder

xiuqhou committed:
Add layer_norm_eps param

xiuqhou committed:
Update src/transformers/models/relation_detr/modeling_relation_detr.py

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

xiuqhou committed:
Update src/transformers/models/relation_detr/modeling_relation_detr.py

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

xiuqhou committed:
Remove empty line

xiuqhou committed:
Update examples

xiuqhou committed:
Remove masks in relation_detr tester

xiuqhou committed:
Rename embed_dim to d_model for consistency

xiuqhou committed:
Add more type hints

xiuqhou committed:
Update docstring

xiuqhou committed:
Add backbone_features_format check

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2871 @@
+        self.post_layer_norm = post_layer_norm
+        if self.post_layer_norm:
+            self.norms = nn.ModuleList([nn.LayerNorm(channel) for channel in in_channels])
+
```
It seems that nn.ModuleList has been used. Is there anything that needs to be improved or have I misunderstood?🤔

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2871 @@
+
+
+# Copied from transformers.models.deformable_detr.modeling_deformable_detr.inverse_sigmoid
+def inverse_sigmoid(x, eps=1e-5):
```
It seems that `copied from` has been added

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2871 @@
+    """
+
+    def __init__(self, config: RelationDetrConfig):
+        super().__init__()
```
Looks like `config` has been used

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2871 @@
+            )
+
+        # expand attention_mask
+        if attention_mask is not None:
+            if attention_mask.dtype == torch.bool:
+                float_attention_mask = torch.zeros_like(attention_mask, dtype=hidden_states.dtype)
+                float_attention_mask.masked_fill_(~attention_mask, float("-inf"))
+                attention_mask = float_attention_mask
+
+            if attention_mask.dim() == 2:
+                if attention_mask.size() != (target_len, source_len):
+                    raise ValueError(
+                        f"2D-Attention mask should be of size {(target_len, source_len)}, but is {attention_mask.size()}"
+                    )
+                # [seq_len, seq_len] -> [batch_size, self.num_heads, target_seq_len, source_seq_len]
+                attention_mask = attention_mask.expand(batch_size, self.num_heads, *attention_mask.size())
+            elif attention_mask.dim() == 3:
+                if attention_mask.size() != (batch_size * self.num_heads, target_len, source_len):
+                    raise ValueError(
+                        f"3D-Attention mask should be of size {(batch_size, self.num_heads, target_len, source_len)}, but"
+                        f" is {attention_mask.size()}"
+                    )
+                attention_mask = attention_mask.view(batch_size, self.num_heads, target_len, source_len)
+            else:
+                raise ValueError(
+                    f"Attention mask should be of size {(target_len, source_len)} of {(batch_size * self.num_heads, target_len, source_len)}, but is"
+                    f" {attention_mask.size()}"
+                )
```
I think we can not prepare it in advance. Because `attention_mask` has different values and dims in different branches and layers of Relation-DETR. This is where relation-detr differs from other models, it adds position relation to `attention_mask`

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2871 @@
+    def with_pos_embed(self, tensor: torch.Tensor, position_embeddings: Optional[Tensor]):
+        return tensor if position_embeddings is None else tensor + position_embeddings
+
+    def forward(
```
Yes, `MultiheadAttention` here is different from existing implementations. It support `attention_mask` with shape of `(batch_size, num_heads, target_len, source_len)`, which means that each header has a different mask. But existing attention implementations in other DETR models only support `(batch_size, 1, target_len, source_len)`.

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2871 @@
+        gt_boxes_list: List[torch.FloatTensor],
+        return_dict: Optional[bool] = None,
+    ):
+        """
+        Args:
+            gt_labels_list (`List[torch.LongTensor]` of len `(batch_size,)`):
+                Classification labels for images in a batch.
+            gt_boxes_list (`List[torch.FloatTensor]` of len `(batch_size,)`):
+                Ground truth bounding boxes for images in a batch, in format `(center_x, center_y, width, height)`.
+            return_dict (`bool`, *optional*):
+                Whether or not to return a [`~file_utils.ModelOutput`] instead of a plain tuple.
+        """
```
Hi, I'm sorry I didn't find out where the docstrings is wrong. Can you give some more details, and I will fix them accordingly.

xiuqhou commented:
Hi @qubvel , thank you very much for your careful review! 🤗 I have updated the code accordingly, and left some responses in your comments to clarify some questions. Please let me know if there is anything that needs to be improved.

qubvel commented:
Hi @xiuqhou, Happy New Year! Thanks for addressing the comments, and sorry for the delay. I'm going to review it right now. 🤗

qubvel commented on src/transformers/models/relation_detr/image_processing_relation_detr.py:
```diff
@@ -0,0 +1,1100 @@
+            images = prepared_images
+            annotations = prepared_annotations
+            del prepared_images, prepared_annotations
+
+        # transformations
+        if do_resize:
+            if annotations is not None:
+                resized_images, resized_annotations = [], []
+                for image, target in zip(images, annotations):
+                    orig_size = get_image_size(image, input_data_format)
+                    resized_image = self.resize(
+                        image, size=size, resample=resample, input_data_format=input_data_format
+                    )
+                    resized_annotation = self.resize_annotation(
+                        target, orig_size, get_image_size(resized_image, input_data_format)
+                    )
+                    resized_images.append(resized_image)
+                    resized_annotations.append(resized_annotation)
+                images = resized_images
+                annotations = resized_annotations
+                del resized_images, resized_annotations
+            else:
+                images = [
+                    self.resize(image, size=size, resample=resample, input_data_format=input_data_format)
+                    for image in images
+                ]
+
+        if do_rescale:
+            images = [self.rescale(image, rescale_factor, input_data_format=input_data_format) for image in images]
+
+        if do_normalize:
+            images = [
+                self.normalize(image, image_mean, image_std, input_data_format=input_data_format) for image in images
+            ]
+
+        if do_convert_annotations and annotations is not None:
+            annotations = [
+                self.normalize_annotation(annotation, get_image_size(image, input_data_format))
+                for annotation, image in zip(annotations, images)
+            ]
```
Lets put this toa  single loop:

```python

processed_images = []
processed_annotations = []

for i, image in enumerate(images):
    annotation = annotations[i] if annotations is not None else None

    if resize:
        image = self.resize(...)
        if annotation is not None:
            annotation = ...
    ...

    processed_images.append(image)
    if annotation is not None:
        processed_annotations.append(annotation)
```

qubvel commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2889 @@
+            nn.init.constant_(module.reference_points.bias.data, 0.0)
+        if hasattr(module, "level_embed"):
+            nn.init.normal_(module.level_embed)
+
+
```
We probably need head initialization for finetuning? see rt-detr's `_init_weights` for example

qubvel reviewed (commented):
Thanks, I did another round of review! It looks clean. There are mostly small comments on my side, please see them below. Next time gonna try fine-tuning and if everything good pass it to core-maintainers approval to be merged 🤗

xiuqhou committed:
raise ValueError for `backbone_feature_format`

xiuqhou committed:
Remove is_timm_available() in relation_detr

xiuqhou committed:
Update copyright year to 2025

xiuqhou committed:
Add copied from in test_modeling_relation_detr

xiuqhou committed:
Add initialization for head of relation_detr

stevhliu commented on docs/source/en/model_doc/relation_detr.md:
```diff
@@ -0,0 +1,102 @@
+
+The Relation-DETR model was proposed in [Relation DETR: Exploring Explicit Position Relation Prior for Object Detection](https://arxiv.org/abs/2407.11699v1) by Xiuquan Hou, Meiqin Liu, Senlin Zhang, Ping Wei, Badong Chen, Xuguang Lan.
+
+Relation DETR is an object detection model that incorporates explicit position relation priors to enhance detection performance. By leveraging the strengths of transformer architectures while integrating spatial relation modeling, Relation DETR achieves superior detection accuracy and fast convergence. This innovative design not only enhances the model's capability to capture complex object inter ...
```
```suggestion
Relation-DETR is an object detection model that incorporates explicit position relation priors to enhance detection performance. By leveraging the strengths of transformer architectures while integrating spatial relation modeling, Relation-DETR achieves superior detection accuracy and fast convergence. This innovative design not only enhances the model's capability to capture complex object interactions but also ensures rapid convergence during training, making it an efficient and high-performance solution for object detection tasks.
```

stevhliu commented on docs/source/en/model_doc/relation_detr.md:
```diff
@@ -0,0 +1,102 @@
+
+## Usage tips
+
+Initially, an image is processed using a pre-trained convolutional neural network, specifically a Resnet variant as referenced in the original code. This network extracts features from the final three layers of the architecture. Following this, a transformer encoder is employed to convert the multi-scale features into a sequential array of image features. Then, a decoder, equipped with auxiliary  ...
```
```suggestion
Initially, an image is processed with a ResNet variant as referenced in the original code. This network extracts features from the final three layers of the architecture. Following this, a transformer encoder is employed to convert the multi-scale features into a sequential array of image features. Then, a decoder, equipped with auxiliary prediction heads and position relation encoders is used to refine the object queries. This process facilitates the direct generation of bounding boxes, eliminating the need for any additional post-processing to acquire the logits and coordinates for the bounding boxes.
```

stevhliu reviewed (commented):
Thanks for the docs!

xiuqhou committed:
Update docs/source/en/model_doc/relation_detr.md

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

xiuqhou committed:
Update docs/source/en/model_doc/relation_detr.md

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

xiuqhou committed:
Fix _init_weights in RelationDetrPretrainedModel

xiuqhou committed:
Update test for relation_detr

xiuqhou committed:
Update src/transformers/models/relation_detr/modeling_relation_detr.py

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

xiuqhou committed:
Put transformation in a single loop

xiuqhou committed:
Update order of ModelOutput

xiuqhou committed:
Update src/transformers/models/relation_detr/modeling_relation_detr.py

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

xiuqhou committed:
Move post_layer_norm to RelationDetrConvEncoder

xiuqhou committed:
Use compile_compatible_lru_cache

xiuqhou committed:
use error raises instead of assert

xiuqhou committed:
Update src/transformers/models/relation_detr/modeling_relation_detr.py

Co-authored-by: Pavel Iakubovskii <qubvel@gmail.com>

xiuqhou committed:
CHECKPOINT in test relation_detr

xiuqhou committed:
Remove mask annotation test for relation_detr

ArthurZucker commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2927 @@
+        self.dropout = config.dropout
+        self.layers = nn.ModuleList([RelationDetrEncoderLayer(config) for _ in range(config.encoder_layers)])
+
+        self.memory_fusion = nn.Sequential(
+            nn.Linear(config.d_model * (config.encoder_layers + 1), config.d_model),
+            nn.ReLU(inplace=True),
+            nn.Linear(config.d_model, config.d_model),
+            nn.LayerNorm(config.d_model, eps=config.layer_norm_eps),
+        )
```
really not a fan of Sequential, let's use Module list

xiuqhou replied:
has replace it with nn.ModuleList

ArthurZucker commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2927 @@
+    return pos_embed
+
+
+def compile_compatible_lru_cache(*lru_args, **lru_kwargs):
```
missing copied from

xiuqhou replied:
I found compile_compatible_lru_cache does not support a separate function outside a class. To ensure the correct compile, I just remove lru_cache here. It doesn't matter.

ArthurZucker commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2927 @@
+    return pos_res
+
+
+def get_sine_pos_embed(
```
I'd rather we use what we have In llama which is pretty standard 😉

xiuqhou replied:
Hi! I found pos_embed in llama does not expose the param `scale` and does not support exchanging x and y dimension for an image, which are needed in Relation-DETR. So I just kept the current implementation. If there is better way to increase this, please let me know.

ArthurZucker commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2927 @@
+        Returns:
+            List of mapped torch.FloatTensor
+        """
+        assert len(inputs) <= len(self.in_channels)
```
no asserts please

xiuqhou replied:
has replaced it with `raise`

ArthurZucker commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2927 @@
+
+
+class RelationDetrChannelMapper(nn.Module):
+    def __init__(
```
missing a bit of doc here, what is this used for?

xiuqhou replied:
has updated the doc here

ArthurZucker commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2927 @@
+        self.encoder_class_head = nn.Linear(config.d_model, config.num_labels)
+        self.hybrid_class_head = nn.Linear(config.d_model, config.num_labels)
+
+        self.encoder_bbox_head = RelationDetrMLPPredictionHead(config.d_model, config.d_model, 4, 3)
+        self.hybrid_bbox_head = RelationDetrMLPPredictionHead(config.d_model, config.d_model, 4, 3)
+
```
if `4` and `3` are hardcoded why need them at all!

xiuqhou replied:
Thanks for the comment! `MLPPredictionHead` has been also used in other places. In `ref_point_head` and `query_scale`, they are not hardcoded. So we need to leave them as parameters instead fixed in the module.

ArthurZucker reviewed (commented):
Thanks for the PR! 🔥 getting there 🤗

xiuqhou committed:
Rename enc_output to encoder_output

xiuqhou committed:
Make init_weight in _init_weights

xiuqhou committed:
Fix `Copied` in test_image_processing_relation_detr

xiuqhou committed:
Update examples in relation-detr

xiuqhou committed:
Make get_reference_points a separate function

xiuqhou committed:
Rename var names

xiuqhou committed:
Re-order all modules according to their usage

xiuqhou committed:
Merge GenerateDN and GenerateCDN

xiuqhou committed:
Add _no_split_modules

xiuqhou committed:
Fix __init__ in GenerateCDNQueries

xiuqhou committed:
Merge branch 'main' into add_relation_detr

xiuqhou committed:
Refactor image_processing_relation_detr_fast

xiuqhou committed:
Fix order in dummy_torchvision_objects

xiuqhou committed:
Fix None annotation processing

xiuqhou committed:
Add usage tips for pipeline

xiuqhou committed:
Remove unused property in RelationDetrConfig

xiuqhou committed:
Refactor get_dim_t and assert in SinePositionEmbedding

xiuqhou committed:
Use `torch.nn.MultiheadAttention` as self_attn in decoder

xiuqhou committed:
Fix last_hidden_state order in test

xiuqhou committed:
Use ModuleList to replace Sequential

xiuqhou committed:
Refactor compile_compatible_method_lru_cache

xiuqhou committed:
Replace assert with raise Exception

xiuqhou committed:
Change Copied from to Modified from for test_retain_grad_hidden_states_attentions

xiuqhou committed:
Remove compile_compatible_method_lru_cache get_dim_t

xiuqhou committed:
Merge branch 'main' into add_relation_detr

xiuqhou committed:
Refactor import order

xiuqhou committed:
Refactor _import_structure order

xiuqhou committed:
Make get_dim_t in_place to reduce memory

xiuqhou committed:
Fix wrong device for pos_embed

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2889 @@
+"""
+
+
+class RelationDetrEncoder(RelationDetrPreTrainedModel):
```
I think yes. RelationDetrEncoder contains `MultiscaleDeformableAttention`, so we need to subclass from `PreTrainedModel` to correctly initialize its weight.

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2889 @@
+        """
+        Get reference points for each feature map. Used in decoder.
+
+        Args:
+            spatial_shapes (`torch.LongTensor` of shape `(num_feature_levels, 2)`):
+                Spatial shapes of each feature map.
+            valid_ratios (`torch.FloatTensor` of shape `(batch_size, num_feature_levels, 2)`):
```
Hi! I just checked the type of `spatial_shapes` and found that it is really `LongTensor`, similar to that in other DETR models in `transformers`.

xiuqhou committed:
Update type_hint for generate_cdn

xiuqhou commented on docs/source/en/model_doc/relation_detr.md:
````diff
@@ -0,0 +1,102 @@
+remote: 0.90 [333.09, 76.71, 369.77, 187.4]
+couch: 0.90 [0.45, 0.53, 640.44, 475.54]
+```
+
````
Yes! I have checked the inference through pipeline `object-detection` and updated docs in `Usage tip` of docs.

xiuqhou commented on src/transformers/models/relation_detr/configuration_relation_detr.py:
```diff
@@ -0,0 +1,304 @@
+
+    @property
+    def hidden_size(self) -> int:
+        return self.d_model
```
Thanks! has removed them

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2927 @@
+        self.backbone_features_format = config.backbone_features_format
+        self.post_layer_norm = config.backbone_post_layer_norm
+
+        if self.post_layer_norm:
```
Only checkpoints with `FocalNet` backbone use this. Other checkpoints (with `ResNet` and `Swin` backbones) do not use this.

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2927 @@
+        self.scale = config.sin_cos_scale
+        self.eps = 1e-6
+        self.offset = config.sin_cos_offset
+        assert (
```
Has updated

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2927 @@
+        if isinstance(self.temperature, int):
+            dim_t = get_dim_t(self.embedding_dim, self.temperature, device)
+            return dim_t, dim_t
+        return (get_dim_t(self.embedding_dim, t, device) for t in self.temperature)
```
has moved this in __init__ for only once calculation

xiuqhou commented on src/transformers/models/relation_detr/modeling_relation_detr.py:
```diff
@@ -0,0 +1,2927 @@
+    def forward(self, pixel_mask: torch.Tensor) -> torch.Tensor:
+        y_embed = pixel_mask.cumsum(1)
+        x_embed = pixel_mask.cumsum(2)
+        if self.normalize:
```
All of released checkpoints use it. The `normalize` parameter is kept only to be compatible with other DETR models. We can also removed it.

xiuqhou committed:
Merge branch 'main' into add_relation_detr

xiuqhou commented on src/transformers/models/relation_detr/image_processing_relation_detr.py:
Thanks for the comment! I have tried the `modular_transformer` but it failed. There may be a bug when converting a file with local imports and `return`. I have opened an issue to discuss this https://github.com/huggingface/transformers/issues/36208

xiuqhou committed:
Merge branch 'main' into add_relation_detr

xiuqhou committed:
Add more type hint

xiuqhou commented:
@ArthurZucker @qubvel @stevhliu Sorry for the late reply! Thanks for your detailed review and suggestions. I have pushed commits according to the reviews except https://github.com/huggingface/transformers/pull/34900#discussion_r1930560623 about the modular style. 
I have finished converting the code to the modular style but met a bug https://github.com/huggingface/transformers/issues/36208 in `utils/modular_model_converter.py`.
Once it is fixed, I will commit this part of code.

CI errors seem not related to changes.

qubvel commented:
Hey @xiuqhou thanks for working on the PR and opening the issue!

I think we can use the following workaround while the issue is not fixed.
Just override the function in modular file to avoid local imports (we can leave numpy and torch only)

```pyhton
def get_numpy_to_framework_fn(arr) -> Callable:
    """
    Returns a function that converts a numpy array to the framework of the input array.

    Args:
        arr (`np.ndarray`): The array to convert.
    """
    if isinstance(arr, np.ndarray):
        return np.array
    elif is_torch_available() and is_torch_tensor(arr):
        return torch.tensor
    else:
        raise RuntimeError("...")
```

Let me know if that works for you!

qubvel commented:
Oh, it might be a redundant function that was copied from one model to another 😄 In that case we can open a separate PR to remove it everywhere

daniel-bogdoll commented:
Hey @xiuqhou , thanks so much for your work so far! Since this will be (most likely) the new best DETR model on HF, I wanted to ask if you think that a merge is likely in the upcoming months? I saw that the issue you mentioned https://github.com/huggingface/transformers/issues/36208 was fixed by https://github.com/huggingface/transformers/pull/36279

xiuqhou committed:
Merge branch 'main' into add_relation_detr

xiuqhou committed:
Code style in loss_utils

xiuqhou committed:
Add modular_relation_detr

xiuqhou committed:
Update image_processing_relation_detr

xiuqhou committed:
Remove GenerationTesterMixin for relation_detr

xiuqhou committed:
Update modular_relation_detr

xiuqhou commented:
Hi @daniel-bogdoll Sorry for the late reply! As https://github.com/huggingface/transformers/issues/36208 has been fixed, I have committed code about the modular style for relation-detr. Now I think the PR is fully prepared for review or merge. Please let me know if there are any remaining improvements needed before merging.

ArthurZucker commented on src/transformers/models/relation_detr/modular_relation_detr.py:
```diff
@@ -0,0 +1,637 @@
+                The size to which the height and width of the images should be divisible by. If set, the images will
+                be padded to the nearest multiple of `size_divisor`.
+        """
+        do_resize = self.do_resize if do_resize is None else do_resize
+        size = self.size if size is None else size
+        size = get_size_dict(size=size, default_to_square=False)
+        resample = self.resample if resample is None else resample
+        do_rescale = self.do_rescale if do_rescale is None else do_rescale
+        rescale_factor = self.rescale_factor if rescale_factor is None else rescale_factor
+        do_normalize = self.do_normalize if do_normalize is None else do_normalize
+        image_mean = self.image_mean if image_mean is None else image_mean
+        image_std = self.image_std if image_std is None else image_std
+        do_convert_annotations = (
+            self.do_convert_annotations if do_convert_annotations is None else do_convert_annotations
+        )
+        do_pad = self.do_pad if do_pad is None else do_pad
+        pad_size = self.pad_size if pad_size is None else pad_size
+        format = self.format if format is None else format
```
cc @yonigozlan we don't need these anymore do we?

ArthurZucker reviewed (commented):
Great work isolating the differences! 🤗 
Need a comment from @yonigozlan but good otherwise!

vncntcmnn commented:
Hey,

Really promising model and I was wondering if there are any updates on getting this PR merged ? 👀

No rush, just curious about the status. 

Thanks !
